### ISSUE TITLE
Native MessageBox API

### ISSUE THREAD

**Issue by zezba9000** [AvaloniaUI/Avalonia#670]
It would be great to have a native Message box API. Win32 on Windows and GTK on Linux / OSX.

**Comment by nc4rrillo** [comment]
I think we should take inspiration from UWP and implement [MessageDialog](https://msdn.microsoft.com/library/windows/apps/xaml/windows.ui.popups.messagedialog.aspx) and [ContentDialog](https://msdn.microsoft.com/library/windows/apps/windows.ui.xaml.controls.contentdialog.aspx)

**Comment by danwalmsley** [comment]
+1 for @nc4rrillo

**Comment by danwalmsley** [comment]
will require #624 to also be completed.

**Comment by TherionIM** [comment]
I know it's a workaround but what about this: https://habr.com/en/post/454386/

**Comment by KevinEady** [comment]
Hi @danwalmsley ,

I see that https://github.com/AvaloniaUI/Avalonia/issues/624 was closed... Is this workable now?

Thanks, Kevin

**Comment by danwalmsley** [comment]
Definitely possible now

**Comment by grokys** [comment]
Given that this issue is about *native* message boxes, I've done a bit of research into the various native message box APIs in the platforms we support:

# Native Message Boxes

## Win32

### MessageBox

`MessageBox` shows a simple message box. It has the following elements:

- A title
- A text message
- A fixed choice of buttons (localized):
  - OK
  - OK, Cancel
  - Yes, No
  - Yes, No, Cancel
  - Abort, Retry, Ignore
  - Cancel, Try Again, Continue
  - An optional Help button an be added to each of the above
- An icon:
  - Exclamation,
  - Warning
  - Information/Asterisk
  - Question
  - Stop/Error/Hand
  
This is the API used by WPF's `MessageBox`.
  
### TaskDialog

`TaskDialog` shows a "vista"-style messagebox. It has the following elements:

- A title
- A message ("main instruction")d
- Content (additional text that appears below the main instruction, in a smaller font)
- Any combination of buttons (localized):
  - OK
  - Yes
  - No
  - Cancel
  - Retry
  - Close
- An icon:
  - Error
  - Information
  - Shield
  - Warning

There is a proposal to add an API for this to WinForms: https://github.com/dotnet/winforms/issues/146

### TaskDialogIndirect
 
`TaskDialogIndirect` shows an advanced "vista"-style messagebox. It has the elements from `TaskDialog` plus:

- A set of buttons that can be:
  - Radio buttons
  - Command links
- A verification checkbox
- An expander to show additional information
- An arbirary footer icon
- A footer message
- Hyperlinks in the content/expander/footer
- A progress bar

## OSX

`NSAlert` shows a message box. It has the following elements:

- A message
- Informative text
- A type
  - Critical
  - Informational/Warning
- A custom icon
- An optional help button
- Arbitrary buttons
- A supression checkbox
- An accessory view containing arbitrary content

Localization of buttons?

## X11

On X11 message boxes are provided by the toolkit so we would have to implement our own managed message boxes, but for context here's what GTK and QT's message box APIs provide.

### GTK

`gtk_message_dialog_new` is simply a convenince function for creating simple dialogs. It can be modified arbirarily, but the base API has the following elements:

- A title
- A message
- Secondary text
- An icon
  - Information
  - Warning
  - Question
  - Error
  - Other (No icon)
- Buttons:
  - OK (discouraged)
  - Close
  - Cancel
  - Yes, No (discouraged)
  - OK, Cancel (discouraged)

### QT

`QMessageBox` shows a message box. It has the following elements:

- Primary text (plain or rich)
- Informative text (plain or rich)
- Optional detailed text
- An icon
  - Question
  - Information
  - Warning
  - Critical
  - A custom bitmap
- Buttons:
  - OK
  - Open
  - Save
  - Cancel
  - Close
  - Discard
  - Apply
  - Reset
  - RestoreDefaults
  - Help
  - Save All
  - Yes
  - Yes to All
  - No
  - No to All
  - Abort
  - Retry
  - Ignore
  
Two APIs for using QMessageBox are provided, the property-based API, and the static functions. Calling one of the static functions is the simpler approach, but it is less flexible than using the property-based API, and the result is less informative. Using the property-based API is recommended.

## Android

`AlertDialog` shows a message box. It has the following elements:

- Title
- Content area (can be text, a list or other custom layout)
- Up to 3 buttons with arbitrary text

## iOS

`UIAlertController` shows a message box. It has the following elements:

- A title
- A message
- A style
  - Action sheet (lays out actions vertically)
  - Alert
- Arbitrary actions (buttons)

**Comment by grokys** [comment]
If we were to implement a native MessageBox API, I can see 3 choices:

- Implement a common subset. This would appear to be:
  - Title (implemented on all platforms except OSX - could just not display it there)
  - Message in plain text
  - Win32 `MessageBox` button choices if we use that API, or a arbitrary buttons if we use `TaskDialog`
    - Does `TaskDialog` look out of place in Win10? Does that matter?
  - "Content" text could be displayed using the relevant feature on platforms that support it, and appended to the message on platforms that don't?
- Have some sort of capabilities query
  - Would somewhat overcomplicate just showing a message box in a cross-platform manner
- Just provide the bare minimum somewhat like JS:
  -  `alert` displays a message and an OK button
  - `confirm` displays a mesage and OK, Cancel buttons

**Comment by kekekeks** [comment]
Note that we need some kind of localized text database anyway since we have default context menus textboxes.

**Comment by grokys** [comment]
Discussion about a new `TaskDialog` API in WinUI: https://github.com/microsoft/microsoft-ui-xaml/issues/2313 - though this is only really a "native" MessageBox in the sense that MS would be shipping it.

**Comment by mysteryx93** [comment]
In terms of UI thread lock, I just worked around this issue in WPF with this simple function.

Can now run sync UI methods in a fully asynchronous way and await!

    /// <summary>
    /// Runs a synchronous action asynchronously on the UI thread.
    /// </summary>
    /// <param name="window">Any window to get the dispatcher from.</param>
    /// <param name="action">The action to run asynchronously.</param>
    /// <typeparam name="T">The return type of the action.</typeparam>
    /// <returns>The result of the action.</returns>
    public static Task<T> RunUiAsync<T>(this Window window, Func<T> action)
    {
        if (window == null) throw new ArgumentNullException(nameof(window));
        TaskCompletionSource<T> completion = new();
        window.Dispatcher.BeginInvoke(new Action(() => completion.SetResult(action())));
        return completion.Task;
    }

**Comment by mysteryx93** [comment]
In the [Avalonia port of MvvmDialogs](https://github.com/mysteryx93/mvvm-dialogs), I need to have a common MessageBox API to share between WPF, Avalonia, UWP, WinUI3, Blazor or anything else (because the ViewModel shouldn't need to know which UI it's running on)

What we first need is a basic MessageBox with standard features which should be simple enough to implement.

Later on, you could add an ExtendedMessageBox in Avalonia with extra features, which will take more time and it will be harder to come with a stable cross-platform API for.

ExtendedMessageBox extension could later be added to MvvmDialogs with support for [WPF/Ookii](https://github.com/ookii-dialogs/ookii-dialogs-wpf), and for Avalonia if that's added. But it's really difficult for these extended feature settings to be platform-agnostic.

My vote is that extended features do not need to be part of the core library but can be added through a separate project, but basic MessageBox features DO need to be part of the core library.

**Comment by Anequit** [comment]
NEED this

**Comment by DamianSuess** [comment]
> NEED this

@Anequit, while we wait for integration of this feature, have you checked out [MessageBox.Avalonia](https://github.com/AvaloniaCommunity/MessageBox.Avalonia)?

**Comment by Anequit** [comment]
> > NEED this
> 
> @Anequit, while we wait for integration of this feature, have you checked out [MessageBox.Avalonia](https://github.com/AvaloniaCommunity/MessageBox.Avalonia)?

Yes I have, but I'm not a huge fan of it personally. I think for the time being I'll opt for an alert popup instead of a message box

**Comment by DmitriyYukhanov** [comment]
> have you checked out [MessageBox.Avalonia](https://github.com/AvaloniaCommunity/MessageBox.Avalonia)?

I've just gave it a spin, and it's not resizing vertically for a long texts:
https://github.com/AvaloniaCommunity/MessageBox.Avalonia/issues/104

It also has no API docs for `MessageBoxStandardParams` etc., looks like something unpolished / buggy to me =\

Would be happy to see out-of-the-box platform-agnostic solution in Avalonia.

**Comment by mysteryx93** [comment]
Yes it has a lot of unpolished rough edges. It had no option for setting a default button; they added it but the implementation is badly designed and looks more like a hack.

**Comment by jsreynolds** [comment]
I'm having a nice laugh finding out a few hours into coding that I basically have to write my own custom message box to give the user feedback. Incredible.

**Comment by DamianSuess** [comment]
As an option, with Prism.Avalonia you make use of the [DialogService](https://github.com/AvaloniaCommunity/Prism.Avalonia/wiki/Dialog-Service) which provides the ability to make it modal and non-modal with ease. It's pretty much 60 seconds of copy/paste as per the [end-to-end sample](https://github.com/AvaloniaCommunity/Prism.Avalonia/tree/develop/e2e/SampleDialogApp)

**Comment by mysteryx93** [comment]
or using [MvvmDialogs for Avalonia](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs)

**Comment by kelvin-martin** [comment]
Simple MessageBox should be higher on the implementation priority list. Cross Platform UI framework without a MessageBox!!

**Comment by sb-viktor** [comment]
This is very necessary and without it it is very inconvenient to develop software!!!
I wonder why this feature is still not available natively under Avalonia?

**Comment by CriticalFlaw** [comment]
Agreed. This feels like a weird omission. MessageBoxes are not uncommon.

**Comment by CriticalFlaw** [comment]
Agreed. This feels like a weird omission. MessageBoxes are not uncommon.

**Comment by vpenades** [comment]
At the very least, a mechanism for basic error/alert notification with basic functionality should be considered.

Something that could rely on native implementations like message boxes or toasts.... or fallback to in built popups.

But it's a pain to go through hoops whenever a command execution throws an expected exception

**Comment by zezba9000** [comment]
The reason message boxes are nice is if you're trying to show something when the UI fails to render.

However for portability reasons, its best practice imo to actually just have a MessageBox overlay. This allows any OS or environment to render messages even when it doesn't support it. Like on console systems or maybe some other mobile platform. You can make a custom one of course pretty easy but having a message overlay system built in wouldn't be bad either.

**Comment by CastelloBrancoTecnologia** [comment]
Please see my implementation at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox 

If it will be usefull I will open an pr to you

**Comment by zezba9000** [comment]
> Please see my implementation at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox
> 
> If it will be usefull I will open an pr to you

You should consider adding Native-Platform Message Boxes. So for Android, iOS, Win32 , SDL for Linux etc.
That way if the app has an unhandled exception and the app can't display anything, you can at least notify the user with a system dialog.

**Comment by vpenades** [comment]
I think there's two completely separate tasks here, and IMHO, I beleve both should go in separate libraries.

Native message boxes are meant to display notifications, warnings and errors, nothing more, nothing less, and as @zezba9000 mentioned, by being native they may be able to show in cases in which Avalonia may not be responsive.

General purpose dialog boxes are a completely different thing meant for presenting forms or other media content to the user.

**Comment by CastelloBrancoTecnologia** [comment]
> > Please see my implementation at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox
> > If it will be usefull I will open an pr to you
> 
> You should consider adding Native-Platform Message Boxes. So for Android, iOS, Win32 , SDL for Linux etc. That way if the app has an unhandled exception and the app can't display anything, you can at least notify the user with a my messabox works before avalonia is avaliable, ie before mainwindow is created on app.cs framework initialization complete.

The only things it not works is in program.cs before avalonia builder has worked. 

Also are there any way to detect if avalonia isn't avaliable? If Yes i can put an fallback to native os messagebox in my code.

**Comment by zezba9000** [comment]
> > > Please see my implementation at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox
> > > If it will be usefull I will open an pr to you
> > 
> > 
> > You should consider adding Native-Platform Message Boxes. So for Android, iOS, Win32 , SDL for Linux etc. That way if the app has an unhandled exception and the app can't display anything, you can at least notify the user with a
> 
> my messabox works before avalonia is avaliable, ie before mainwindow is created on app.cs framework initialization complete.
> 
> The only things it not works is in program.cs before avalonia builder has worked.
> 
> Also are there any way to detect if avalonia isn't avaliable? If Yes i can put an fallback to native os messagebox in my code.

The only way to catch 100% of issues is probably to try/catch around "BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);"
If you wanted to support native MessaageBoxes in your API, you shouldn't auto detect this and instead have a very limited API subset or not part of the main API people can manually invoke letting them choose based on how their app works.

A couple reasons for this are, #1 native message boxes are much more limited or maybe need some extra options specific to them only and #2 you end up causing issues overengineering solutions that will never solve all edge cases. So keep it simple.

I've added Native Message boxes on literally all kinds of platforms (even BB10). I would do this myself but I've got other projects atm. To make it easy in C#, just use the dotnet workloads etc. Then on Linux, just wrap what you need for SDL (or I can show you how to make a native Wayland or X11 dialog if you wanted). https://wiki.libsdl.org/SDL2/SDL_ShowSimpleMessageBox

**Comment by CastelloBrancoTecnologia** [comment]
> > > > Please see my implementation at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox
> > > > If it will be usefull I will open an pr to you
> > > 
> > > 
> > > You should consider adding Native-Platform Message Boxes. So for Android, iOS, Win32 , SDL for Linux etc. That way if the app has an unhandled exception and the app can't display anything, you can at least notify the user with a
> > 
> > 
> > my messabox works before avalonia is avaliable, ie before mainwindow is created on app.cs framework initialization complete.
> > The only things it not works is in program.cs before avalonia builder has worked.
> > Also are there any way to detect if avalonia isn't avaliable? If Yes i can put an fallback to native os messagebox in my code.
> 
> The only way to catch 100% of issues is probably to try/catch around "BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);" If you wanted to support native MessaageBoxes in your API, you shouldn't auto detect this and instead have a very limited API subset or not part of the main API people can manually invoke letting them choose based on how their app works. BUT I would do all this in a new project because people may only need native message boxes and not your other solution. Keep the bloat down.
> 
> A couple reasons for this are, [#1](https://github.com/AvaloniaUI/Avalonia/issues/1) native message boxes are much more limited or maybe need some extra options specific to them only and [#2](https://github.com/AvaloniaUI/Avalonia/issues/2) you end up causing issues overengineering solutions that will never solve all edge cases. So keep it simple.
> 
> I've added Native Message boxes on literally all kinds of platforms (even BB10). I would do this myself but I've got other projects atm. To make it easy in C#, just use the dotnet workloads etc. Then on Linux, just wrap what you need for SDL (or I can show you how to make a native Wayland or X11 dialog if you wanted). https://wiki.libsdl.org/SDL2/SDL_ShowSimpleMessageBox
> 
> The other and better options outside dotnet workloads so it doesn't require them is to just invoke the native APIs directly. On Linux using C this is easy. On macOS just make an ObjC component (don't use Swift) and invoke a extern C method that call ObjC code (this is very easy to do)

I implemented an native messagebox for macos, windows and linux using x11 (that is suported by wayland with xwayland that is default)

please see 

https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox/blob/main/AvaloniaMessageBox/AvaloniaMessageBox.cs

Not testes beause is WIP.

**Comment by CastelloBrancoTecnologia** [comment]
Native messagebox implemented as a fallback of avaloniamessagebox.

Please see my code

**Comment by CastelloBrancoTecnologia** [comment]
@kekekeks are there any interest in my implementation?  I want tô contribute it tô avalonia. If Yes i can make an PR or you can get the code directly.

**Comment by zezba9000** [comment]
> [@kekekeks](https://github.com/kekekeks) are there any interest in my implementation? I want tô contribute it tô avalonia. If Yes i can make an PR or you can get the code directly.

I really think you should separate the Native Message Box project from the AvaloniaUI one. Then just make a pull request. Clearly there is interest. I don't run the project but core APIs should be separated like this.

**Comment by CastelloBrancoTecnologia** [comment]
> > [@kekekeks](https://github.com/kekekeks) are there any interest in my implementation? I want tô contribute it tô avalonia. If Yes i can make an PR or you can get the code directly.
> 
> I really think you should separate the Native Message Box project from the AvaloniaUI one. Then just make a pull request. Clearly there is interest. I don't run the project but core APIs should be separated like this.

But it is already separeted from avaloniaui. The project only contains avaloniamessagebox class.

See at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox

I am missing something?

**Comment by zezba9000** [comment]
> > > [@kekekeks](https://github.com/kekekeks) are there any interest in my implementation? I want tô contribute it tô avalonia. If Yes i can make an PR or you can get the code directly.
> > 
> > 
> > I really think you should separate the Native Message Box project from the AvaloniaUI one. Then just make a pull request. Clearly there is interest. I don't run the project but core APIs should be separated like this.
> 
> But it is already separeted from avaloniaui. The project only contains messagebox class and internal nativemessagebox class that are called from messajebox if avalonia isnt avaliable (fallback).
> 
> See at https://github.com/CastelloBrancoTecnologia/AvaloniaMessageBox
> 
> I am missing something?

I see your native message box class. It looks generally good glancing at it but don't make the main API auto fallback to this unless you give an option to disable that. The reason is some Apps may have their own fallback methods and this fallback could be redundant or conflicting.

If you want to make a pull-request you're going to need to fork Avalonia and add this in under Avalonia namespaces (not your own) and go through that process.

**Comment by xmaxrayx** [comment]
just write custom class? especially when everyone has deferent design especially when sync/async come needy if you are going to multi-platform then you should spend time to do that.

idk why dev should waste their time when you can spend couple hours? beside Linux *I think* isn't stable but windows have same API messageboxA and messageboxW for years with same int return value

id rather devs make their control better than doing system warper.

**Comment by zezba9000** [comment]
> just write custom class? especially when everyone has deferent design especially when sync/async come needy if you are going to multi-platform then you should spend time to do that.
> 
> idk why dev should waste their time when you can spend couple hours? beside Linux _I think_ isn't stable but windows have same API messageboxA and messageboxW for years with same int return value
> 
> id rather devs make their control better than doing system warper.

System wrappers are needed to show error message boxes if AvaloniaUI fails to init. On Linux, you can just use SDL. Every platform has this as its needed for crit error messages. Android and iOS apps do the same thing.

**Comment by xmaxrayx** [comment]
> > just write custom class? especially when everyone has deferent design especially when sync/async come needy if you are going to multi-platform then you should spend time to do that.
> > idk why dev should waste their time when you can spend couple hours? beside Linux _I think_ isn't stable but windows have same API messageboxA and messageboxW for years with same int return value
> > id rather devs make their control better than doing system warper.
> 
> System wrappers are needed to show error message boxes if AvaloniaUI fails to init. On Linux, you can just use SDL. Every platform has this as its needed for crit error messages. Android and iOS apps do the same thing.

expect is can be done with your class and warp it or do 3rd party implantation,

Avalnoia team is not that big and their focus should for ui avalonia itself , system warpers is done by other devs bc it doesn't relay on specific farmwork and can be done with c# level

lets not mention how llinux and andriod can be problematic and wasting of time for doing multi fix lmao

**Comment by zezba9000** [comment]
> lets not mention how llinux and andriod can be problematic and wasting of time for doing multi fix lmao

Android is super easy (idk why you think its not). Linux you just use SDL which is also easy. And no I disagree with you, as with your logic here creating a native Window to render AvaloniaUI in shouldn't be 1st party because it called into the native OS stack. This makes no sense. Everything about app dev calls into native OS level features to operate and showing critical errors when an app crashes requires this. Its fundamental and basic apps or games. AvaloniaUI can't render anything if you hit an unhanded exception but could show a native dialog.

Honestly I should just do this myself as its not hard and make a pull request. I've just got other projects and keep putting it off.

**Comment by stevemonaco** [comment]
"Just use SDL" means taking a SDL dependency on Linux which is almost certainly a non-starter for being included in core Avalonia. It doesn't give devs anything much in the case of an actual crash when Avalonia is exiting and the MessageBox (native or otherwise) doesn't have a stable process. If your process is stable, then show an overlay dialog.

**Comment by zezba9000** [comment]
> "Just use SDL" means taking a SDL dependency on Linux which is almost certainly a non-starter for being included in core Avalonia. It doesn't give devs anything much in the case of an actual crash when Avalonia is exiting and the MessageBox (native or otherwise) doesn't have a stable process. If your process is stable, then show an overlay dialog.

Wrong. Please if you have never done Linux dev stop making false claims to its implementation philosophy. Linux and particularity Wayland do not have standards when it comes to native system message boxes. SDL is considered the standard in most cases for stuff like this. Implementing this manually is not like any other platform and going over the reasons why complex.

"If your process is stable"
-- O my dear dear friend... if only magic fairy dusk could fix all problems we wouldn't need roads and could fly.

**Comment by stevemonaco** [comment]
@zezba9000 Explain to me how using SDL for Linux is not taking a dependency on SDL without being rude.

I separate stable vs unstable processes because unstable processes require different treatment. Whether that's a secondary process (native or supplied) to show a dialog or if it's writing a log which gets loaded on the app's next start will depend on the needs. If there are problems where Avalonia or .NET won't launch, then something secondary is necessary.

**Comment by vpenades** [comment]
Regarding SDL on linux, I've reviewed the source code [I linked before](https://github.com/AvaloniaUI/Avalonia/issues/670#issuecomment-2810705742) , it does not use SDL on linux, but it uses GTK , I don't know if this would be an acceptable solution.

**Comment by zezba9000** [comment]
Do people not understand what "Native" Message box means? In the context of Linux, there is no such thing as a "Native" MessageBox without a toolkit like SDL, GTK, KDE etc. Thats how GNU Linux works. I'm not going to argue this topic anymore.

**Comment by TSRBerry** [comment]
I don't understand why people are spamming this issue, arguing whether it's necessary or not.
The issue was assigned to a milestone and a person already, why are people arguing here?

**Comment by xmaxrayx** [comment]
> > lets not mention how llinux and andriod can be problematic and wasting of time for doing multi fix lmao
> 
> Android is super easy (idk why you think its not). Linux you just use SDL which is also easy. And no I disagree with you, as with your logic here creating a native Window to render AvaloniaUI in shouldn't be 1st party because it called into the native OS stack. This makes no sense. Everything about app dev calls into native OS level features to operate and showing critical errors when an app crashes requires this. Its fundamental and basic apps or games. AvaloniaUI can't render anything if you hit an unhanded exception but could show a native dialog.
> 
> Honestly I should just do this myself as its not hard and make a pull request. I've just got other projects and keep putting it off.

you can use other stuff like out print the error in console or log text even you can send the error in clip bored you crush in message box with user base who should be knowledge of using console?


>Android is super easy (idk why you think its not). Linux you just use SDL which is also easy. 

there is no standard in linux all just opinions and popularity usage of said pkg lmao, even you pray that said pkj isn't alerted by pkg maintainer which good luck trusting random ppl on internet do random stuff and may call it for security or whatever

for andriod idk i dont want trust google bc they make new version of android less good and more restricted.

>Everything about app dev calls into native OS level features to operate and showing critical errors when an app crashes requires this. Its fundamental and basic apps or games. AvaloniaUI can't render anything if you hit an unhanded exception but could show a native dialog.

its Linux community issue for having random and spaghetti design also what you gona call? GTK3-2-1 or QT5-4-5-6-3-2-1 or another Desktop that use modified version of GTK? or just render whole window as fake message box with another render? see the issue here? in windows you just call same api for years.
why wasting devs time just for small number of consumers that need too much work for little gain.

lets pass you can finally "call" the messagbox you still have another issue which is culture and non-english text support and good i dont have faith with USA devs or "high level" coders bc its beyond write good code when Propaganda and war is hot these days and theses stuff need a lot of communication and being patience bc string wasn't mature and just being added above old design with secret letters like BOM character and righ to left char in string and you need too more RAM if you want support outside ASCCI char in general.

also wasn't wayland block of key bored hock usage? which needed for IME because there languages that can't be fitted on keybored like most Asinas language

so linux thing should be with 3rd party pkg we have nuget and we dont need everything in avalonia, better for devs save their time for avalonia things, the thing isnt about just good code you need have too much faith with no agreement with linux users.

**Comment by CastelloBrancoTecnologia** [comment]
> > > lets not mention how llinux and andriod can be problematic and wasting of time for doing multi fix lmao
> > 
> > 
> > Android is super easy (idk why you think its not). Linux you just use SDL which is also easy. And no I disagree with you, as with your logic here creating a native Window to render AvaloniaUI in shouldn't be 1st party because it called into the native OS stack. This makes no sense. Everything about app dev calls into native OS level features to operate and showing critical errors when an app crashes requires this. Its fundamental and basic apps or games. AvaloniaUI can't render anything if you hit an unhanded exception but could show a native dialog.
> > Honestly I should just do this myself as its not hard and make a pull request. I've just got other projects and keep putting it off.
> 
> you can use other stuff like out print the error in console or log text even you can send the error in clip bored you crush in message box with user base who should be knowledge of using console?
> 
> > Android is super easy (idk why you think its not). Linux you just use SDL which is also easy.
> 
> there is no standard in linux all just opinions and popularity usage of said pkg lmao, even you pray that said pkj isn't alerted by pkg maintainer which good luck trusting random ppl on internet do random stuff and may call it for security or whatever
> 
> for andriod idk i dont want trust google bc they make new version of android less good and more restricted.
> 
> > Everything about app dev calls into native OS level features to operate and showing critical errors when an app crashes requires this. Its fundamental and basic apps or games. AvaloniaUI can't render anything if you hit an unhanded exception but could show a native dialog.
> 
> its Linux community issue for having random and spaghetti design also what you gona call? GTK3-2-1 or QT5-4-5-6-3-2-1 or another Desktop that use modified version of GTK? or just render whole window as fake message box with another render? see the issue here? in windows you just call same api for years. why wasting devs time just for small number of consumers that need too much work for little gain.
> 
> lets pass you can finally "call" the messagbox you still have another issue which is culture and non-english text support and good i dont have faith with USA devs or "high level" coders bc its beyond write good code when Propaganda and war is hot these days and theses stuff need a lot of communication and being patience bc string wasn't mature and just being added above old design with secret letters like BOM character and righ to left char in string and you need too more RAM if you want support outside ASCCI char in general.
> 
> also wasn't wayland block of key bored hock usage? which needed for IME because there languages that can't be fitted on keybored like most Asinas language
> 
> so linux thing should be with 3rd party pkg we have nuget and we dont need everything in avalonia, better for devs save their time for avalonia things, the thing isnt about just good code you need have too much faith with no agreement with linux users.

On linux you see the 2 last versions of more used flavors of toolkits 

if it installed ( gtk4, gtk3, qt5 or qt4, or x11) use it, fallbacking to sdl. 

Also using pinvoke and dinamic loading to avoid the requirement of include them inside avalonia and avoid dependences.

**Comment by udlose** [comment]
still open after 9 years? this is such a fundamental element of a UI framework!

**Comment by xmaxrayx** [comment]
Linux community will be shocked once we talk about their clipboard and how most libraries support only text lol,

hint linux is the issue not avalonia, no one like warp for different DE in linux unless u have been paid for that
,dont forget x11 and wayland version

**Comment by Athari** [comment]
Apparently, this project existed, before it was archived about 2 weeks ago:

* [NativeMessageBox](https://github.com/wieslawsoltes/NativeMessageBox) by @wieslawsoltes

I don't know what the history behind it is, but the API looks promising. Rather than going for the common subset, he went for a superset.

Here's an excerpt from the ReadMe:

----

### NativeMessageBox

NativeMessageBox is a production-ready native dialog runtime that ships a stable C ABI, high-level .NET 8 wrapper, and first-class tooling for Windows, macOS, Linux, iOS, and Android. The project focuses on predictable behaviour, strong diagnostics, and packaging that fits both managed and native distribution pipelines.

#### Feature Summary

| Feature | Description |
| --- | --- |
| Stable C ABI | `include/native_message_box.h` exposes a forward-compatible ABI with explicit struct sizing, version negotiation, and allocator hooks. |
| Native implementations | Dedicated Windows (Task Dialog / MessageBox), macOS (NSAlert), Linux (GTK 3/4 with zenity fallback), iOS (UIKit), Android (AlertDialog), plus a WebAssembly browser overlay host. |
| Managed .NET 8 wrapper | `NativeMessageBox` NuGet package with source-generated interop, async APIs, configurable host abstraction, and logging hooks. |
| Rich dialog options | Multiple buttons, custom IDs, icons, verification checkboxes, timeouts, secondary content, and optional text/password/combo inputs (platform-dependent). |
| Mobile packaging | Automated AAR (Android) and XCFramework (iOS) outputs with manifests describing ABIs, architectures, and build metadata. |
| Tooling and CI | Cross-platform `build.sh`/`build.ps1`, native tests via `ctest`, managed unit tests, and packaging jobs suitable for CI/CD pipelines. |
| Samples | Avalonia desktop and mobile samples demonstrating host integration, lifecycle management, and advanced dialog scenarios. |
| Documentation | DocFX site under `docs/` covering architecture, API reference, troubleshooting, and platform-specific guidance. |

#### Feature Matrix

| Capability | Windows | macOS | Linux (GTK) | iOS | Android | Web (Browser) |
| --- | --- | --- | --- | --- | --- | --- |
| Multi-button dialogs | Yes (Task Dialog supports 8+) | Yes | Yes | Yes | Partial (AlertDialog: 3 buttons) | Yes |
| Custom button text/IDs | Yes | Yes | Yes | Yes | Yes (first 3 buttons) | Yes |
| Button roles (default/cancel/destructive/help) | Yes | Yes | Partial (default/cancel) | Yes | Partial (positive/negative/neutral) | Yes (primary/cancel/destructive) |
| Standard icons | Yes | Yes | Yes | No (ignored) | No (ignored) | No (style via CSS) |
| Verification checkbox | Yes | Yes | Yes | No | No | Yes |
| Text/password input | No (planned) | Yes | Yes | Yes (text/password) | No | Yes |
| Combo box input | No | Yes | Yes | No | No | Yes |
| Secondary informative/expanded content | Yes | Yes | Yes | No | No | Yes |
| Help links / hyperlinks | Yes (Task Dialog hyperlink events) | Yes (opens via `NSWorkspace`) | Yes (GtkLinkButton) | No | No | No (render text only) |
| Auto-close timeout | Yes | Yes | Yes | Yes | No | Yes |
| Threading requirements | STA enforced for advanced dialogs | Must run on main thread | GTK main loop required | Must be called on main thread | Requires `Activity` on UI thread | Browser main thread (async overlay) |

----

I'd love to see something like this in Avalonia.

**Comment by pyking-wang** [comment]
Ten years have passed, and we are still waiting.

**Comment by rent-a-developer** [comment]
Please implement this.
Even a basic message box with title, content and some buttons (e.g. okay, cancel, abort, yes, no) would already help in many situations.
